src/MacroExpander.js

```diff
@@ -17,114 +17,194 @@ class MacroExpander {
                         throw new ParseError(
-                            "End of input expecting macro argument", topToken);
+                            "Incomplete placeholder at end of macro body",
+                            tok);
+                    }
+                    tok = expansion[--i]; // next token on stack
+                    if (tok.text === "#") { // ## → #
+                        expansion.splice(i + 1, 1); // drop first #
+                    } else if (/^[1-9]$/.test(tok.text)) {
+                        // expansion.splice(i, 2, arg[0], arg[1], …)
+                        // to replace placeholder with the indicated argument.
+                        // TODO: use spread once we move to ES2015
+                        expansion.splice.apply(
+                            expansion,
+                            [i, 2].concat(args[tok.text - 1]));
                     } else {
-                        args[i] = [startOfArg];
+                        throw new ParseError(
+                            "Not a valid argument number",
+                            tok);
                     }
                 }
-                // paste arguments in place of the placeholders
-                expansion = expansion.slice(); // make a shallow copy
-                for (i = expansion.length - 1; i >= 0; --i) {
-                    tok = expansion[i];
-                    if (tok.text === "#") {
-                        if (i === 0) {
-                            throw new ParseError(
-                                "Incomplete placeholder at end of macro body",
-                                tok);
-                        }
-                        tok = expansion[--i]; // next token on stack
-                        if (tok.text === "#") { // ## → #
-                            expansion.splice(i + 1, 1); // drop first #
-                        } else if (/^[1-9]$/.test(tok.text)) {
-                            // expansion.splice(i, 2, arg[0], arg[1], …)
-                            // to replace placeholder with the indicated argument.
-                            // TODO: use spread once we move to ES2015
-                            expansion.splice.apply(
-                                expansion,
-                                [i, 2].concat(args[tok.text - 1]));
-                        } else {
-                            throw new ParseError(
-                                "Not a valid argument number",
-                                tok);
-                        }
-                    }
+            }
+        }
+        // Concatenate expansion onto top of stack.
+        this.stack.push.apply(this.stack, expansion);
+        return expansion;
+    }
+
+    /**
+     * Expand the next token once (ignoring initial spaces like `get`),
+     * without removing anything from the stack, and return the top token
+     * on the stack.  Similar in behavior to TeX's `\expandafter\futurelet`.
```

Comment:
How does it ignore initial spaces if it calls `expandOnce` right away which only ignores trailing spaces?

Reply:
That comment was out-of-date.  Updated.